src/librustc_typeck/check/mod.rs

```diff
@@ -3538,9 +3555,11 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             _ => self.warn_if_unreachable(expr.id, expr.span, "expression")
         }
 
-        // Any expression that produces a value of type `!` must have diverged
-        if ty.is_never() {
-            self.diverges.set(self.diverges.get() | Diverges::Always);
+        // Any expression that produces a value of an uninhabited type must have diverged.
+        if ty.conservative_is_uninhabited() {
+            if ty.is_never() || self.tcx.features().better_divergence_checking {
```

Comment:
seems good 👍 -- in particular, seems like a natural extension of the existing check, and it's not clear that `!` should be special here